benchmarks/trees/train.py

```diff
@@ -46,58 +42,37 @@
 
 class TrainEnsembleAlgorithm(ABC):
     @staticmethod
-    def create(name):  # pylint: disable=too-many-return-statements
+    def create(name, learning_task):  # pylint: disable=too-many-return-statements
         if name == "xgb":
-            return XgbAlgorithm()
+            return XgbAlgorithm(learning_task)
         if name == "lgbm":
-            return LgbmAlgorithm()
+            return LgbmAlgorithm(learning_task)
         if name == "rf":
-            return RandomForestAlgorithm()
+            return RandomForestAlgorithm(learning_task)
         raise ValueError("Unknown algorithm: " + name)
 
-    @staticmethod
-    def get_data(data, start, end):
-        return data[start:end, :] if isinstance(data, np.ndarray) else data.iloc[start:end, :]
-
-    def __init__(self):
+    def __init__(self, learning_task):
         self.params = {}
         self.model = None
         self.predictions = []
+        self.learning_task = learning_task
 
     @abstractmethod
     def fit(self, data, args):
         pass
 
     def test(self, data):
         assert self.model is not None
-        return self.model.predict(data.X_test)
+        return self.model.predict(data)
 
-    def predict(self, model, data, args):
-        batch_size = args.batch_size
-        predict_data = data.X_test
+    def predict(self, model, predict_data, args):
+        if self.learning_task == LearningTask.REGRESSION:
+            predict_fn = model.predict
+        else:
+            predict_fn = model.predict_proba
 
         with Timer() as t:
-            total_size = len(predict_data)
-            iterations = total_size // batch_size
-            iterations += 1 if total_size % batch_size > 0 else 0
-            iterations = max(1, iterations)
-
-            if data.learning_task == LearningTask.CLASSIFICATION:
-                self.predictions = np.empty([total_size, 2], dtype="f4")
-                predict_fn = model.predict_proba
-            if data.learning_task == LearningTask.MULTICLASS_CLASSIFICATION:
-                self.predictions = np.empty([total_size, model.n_classes_], dtype="f4")
-                predict_fn = model.predict_proba
-            if data.learning_task == LearningTask.REGRESSION:
-                self.predictions = np.empty([total_size], dtype="f4")
-                predict_fn = model.predict
-
-            for i in range(0, iterations):
-                start = i * batch_size
-                end = min(start + batch_size, total_size)
-                batch = TrainEnsembleAlgorithm.get_data(predict_data, start, end)
-                self.predictions[start:end] = predict_fn(batch)
-
+            self.predictions = predict_fn(predict_data)
```

Comment:
This change may be controversial, but I strongly believe we shouldn't add batching to rf/xgb/lgm, since they can do prediction in one go. Batching is only needed for HB + some backends that require fixed input shape (TVM), so it is only HB's minor technical concern and shouldn't affect rf/xgb/lgm.

Adding batching also to rf/xgb/lgm for the sake of "apple to apple comparison" makes things considerably slower for them, so I don't think this is a fair benchmark. Moreover, by allowing TVM to compile for a fixed shape, which enables many static optimization, I argue that this benchmark is already not "apple to apple comparsion".

We should compare the best rf/xgb/lgm perf vs the best HB perf, since that's what people care about. If we report rf/xgb/lgm numbers with batch by batch prediction, that wouldn't reflect the numbers people would see in practice.  

@interesaaat @ksaur

Reply:
I don't agree with this. I think we should have both, and benchmark on both. We want to support (and benchmark) batching because we want (1) out of memory computations (which are not supported by sklearn/xgb/lgm); (2) an easy way to benchmark online (single record) inference.

Instead of removing this, we can just use the `batch_benchmark` flag: if enabled we do batching (as the code is set now), if not, we just do inference over the whole dataset.

Reply:
ok I'll remove this change since this is not essential for this PR and it needs further discussion. I can see that for out of memory prediction benchmark use case, having batching in xgb makes sense (assuming `xgb_model.predict(huge_data)` in one go doesn't work). 

But I'd say that is rather special use case - this is not something we should benchmark **by default**. You are essentially saying, by default when you run `python run.py`, which means batch size = 10000, you are benchmarking for the hypothetical, out of core use case where we decide to run batch by batch inference with batch size 10000. Causal users might run `python run.py` (which means out of core benchmark) and get the impression that HB is 5x faster, say, when it is not.

So for the benchmark that runs by default, I think it is better to do batch benchmark using the new option I added in the PR, or benchmark on the whole dataset as one batch. Either way, batching in rf/xgb won't take effect so we get the best result they can achieve. On the HB + TVM side, we can do either batch / no batch, depending on the `batch_size` argument. I believe this is the most fair benchmark and these are the numbers we should report by default.

Reply:
Restored batching for xgb